File: sw/rust/map/src/dyn_coord_map.rs
use tagma_core::Coord;
use tagma_core::DynCoordSpace;

use crate::string_to_coord_path;
use crate::CoordMap;

// ── Internal helpers ──────────────────────────────────────────────────────

fn vec_to_box(v: Vec<u8>) -> Box<[u8]> {
    v.into_boxed_slice()
}

fn box_to_vec(v: &[u8]) -> Vec<u8> {
    v.to_vec()
}

fn box_to_vec_owned(v: Box<[u8]>) -> Vec<u8> {
    v.into_vec()
}

fn coords_to_key_bytes(coords: &[Coord]) -> Vec<u8> {
    coords.iter().map(|c| c.index() as u8).collect()
}

// ── DynCoordMap ────────────────────────────────────────────────────────────

/// A hash-free, collision-free string map backed by [`DynCoordSpace`]
/// with [`ByteWise`](crate::ByteWise) coordinate generation.
///
/// Supports any non-empty string key.  Lookup cost is O(len(key)).
///
/// # Trait implementations
///
/// | Trait | Methods |
/// |-------|---------|
/// | [`CoordMap`] | `insert`, `get`, `remove`, `contains_key` via `&str` |
///
/// # Example
///
/// ```
/// use tagma_map::CoordMap;
/// use tagma_map::dyn_coord_map::DynCoordMap;
///
/// let mut kv = DynCoordMap::new();
/// kv.insert("hello", b"world".to_vec());
/// assert_eq!(kv.get("hello"), Some(b"world".to_vec()));
/// ```
pub struct DynCoordMap {
    space: DynCoordSpace<Box<[u8]>>,
    len: usize,
}

impl DynCoordMap {
    /// Creates an empty dynamic-mode store.
    pub fn new() -> Self {
        DynCoordMap {
            space: DynCoordSpace::new(),
            len: 0,
        }
    }

    /// Looks up a value by a fixed-length CoordPath (converts internally).
    pub fn get_by_coord_path<const N: usize>(
        &self,
        path: &tagma_core::CoordPath<N>,
    ) -> Option<Vec<u8>> {
        let coords: Vec<Coord> = path.coords().to_vec();
        self.space.at(&coords).map(|v| box_to_vec(v.as_ref()))
    }

    /// Returns an iterator over `(key_bytes, value)` pairs.
    ///
    /// Keys are reconstructed from the stored Coord path: each Coord is
    /// converted back to the original byte value.  Order is
    /// depth-first, coordinate-ascending.
    pub fn iter(&self) -> DynCoordMapIter<'_> {
        DynCoordMapIter {
            inner: self.space.iter(),
        }
    }
}

impl CoordMap for DynCoordMap {
    fn len(&self) -> usize {
        self.len
    }

    fn is_empty(&self) -> bool {
        self.len == 0
    }

    fn clear(&mut self) {
        self.space.clear();
        self.len = 0;
    }

    fn insert(&mut self, key: &str, value: Vec<u8>) -> Option<Vec<u8>> {
        if key.is_empty() {
            return None;
        }
        let path = string_to_coord_path(key).unwrap();
        let prev = self.space.place(&path, vec_to_box(value));
        if prev.is_none() {
            self.len += 1;
        }
        prev.map(box_to_vec_owned)
    }

    fn get(&self, key: &str) -> Option<Vec<u8>> {
        if key.is_empty() {
            return None;
        }
        string_to_coord_path(key)
            .and_then(|path| self.space.at(&path))
            .map(|v| box_to_vec(v.as_ref()))
    }

    fn remove(&mut self, key: &str) -> Option<Vec<u8>> {
        if key.is_empty() {
            return None;
        }
        let val = string_to_coord_path(key)
            .and_then(|path| self.space.vacate(&path))
            .map(box_to_vec_owned);
        if val.is_some() {
            self.len = self.len.saturating_sub(1);
        }
        val
    }
}

impl Default for DynCoordMap {
    fn default() -> Self {
        Self::new()
    }
}

// ── Iterator ──────────────────────────────────────────────────────────────

/// An iterator over [`DynCoordMap`] entries, yielding `(key_bytes, value)`.
pub struct DynCoordMapIter<'a> {
    inner: tagma_core::DynIter<'a, Box<[u8]>>,
}

impl<'a> Iterator for DynCoordMapIter<'a> {
    type Item = (Vec<u8>, &'a [u8]);

    fn next(&mut self) -> Option<Self::Item> {
        let (path, val) = self.inner.next()?;
        let key = coords_to_key_bytes(&path);
        Some((key, val.as_ref()))
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.inner.size_hint()
    }
}

File: sw/java/tagma-map/src/main/java/org/ssccs/syntagma/map/DynCoordMap.java
package org.ssccs.syntagma.map;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

import org.ssccs.syntagma.core.Coord;
import org.ssccs.syntagma.core.CoordPath;
import org.ssccs.syntagma.core.DynCoordSpace;

/**
 * A hash-free, collision-free string map backed by {@link DynCoordSpace} with
 * byte-wise coordinate generation.
 *
 * <p>Any non-empty string is a valid key and lookup cost is
 * {@code O(key length)}; the path length equals the UTF-8 byte length of the
 * key. An empty key cannot produce a path and is rejected by every operation
 * without storing anything, mirroring the references, where the rejection is
 * reported with the same empty result as an absent key.
 *
 * <p>Port of the C++ {@code tagma_map::DynCoordMap} in
 * {@code sw/cpp/tagma_map/include/tagma_map/dyn_coord_map.h}; the underlying
 * behavior mirrors the Rust {@code DynCoordMap} in
 * {@code sw/rust/map/src/dyn_coord_map.rs}.
 */
public final class DynCoordMap implements CoordMap, CoordPathLookup {

    private final DynCoordSpace<byte[]> space = new DynCoordSpace<>();
    private int len;

    /** Creates an empty dynamic-mode store. */
    public DynCoordMap() {
    }

    @Override
    public Optional<byte[]> insert(String key, byte[] value) {
        Objects.requireNonNull(key, "key");
        Objects.requireNonNull(value, "value");
        Optional<List<Coord>> path = CoordGen.stringToCoordPath(key);
        if (path.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> previous = space.place(path.orElseThrow(), value.clone());
        if (previous.isEmpty()) {
            len += 1;
        }
        return previous;
    }

    @Override
    public Optional<byte[]> get(String key) {
        Objects.requireNonNull(key, "key");
        Optional<List<Coord>> path = CoordGen.stringToCoordPath(key);
        if (path.isEmpty()) {
            return Optional.empty();
        }
        return space.at(path.orElseThrow()).map(byte[]::clone);
    }

    @Override
    public Optional<byte[]> remove(String key) {
        Objects.requireNonNull(key, "key");
        Optional<List<Coord>> path = CoordGen.stringToCoordPath(key);
        if (path.isEmpty()) {
            return Optional.empty();
        }
        Optional<byte[]> removed = space.vacate(path.orElseThrow()).map(byte[]::clone);
        if (removed.isPresent()) {
            len -= 1;
        }
        return removed;
    }

    @Override
    public Optional<byte[]> getByCoordPath(CoordPath path) {
        Objects.requireNonNull(path, "path");
        return space.at(Arrays.asList(path.coords())).map(byte[]::clone);
    }

    /**
     * All {@code (key, value)} pairs in depth-first coordinate-ascending order.
     * Keys are reconstructed from the stored byte-wise path and decoded as
     * ISO-8859-1, so each character is the stored byte value of the references
     * ({@code std::string} and {@code Vec<u8>}). The references hand out
     * pointers into the store; Java hands out copies, so an iterated value is
     * never shared with the store.
     */
    public List<Map.Entry<String, byte[]>> iter() {
        List<Map.Entry<List<Coord>, byte[]>> entries = space.entries();
        List<Map.Entry<String, byte[]>> out = new ArrayList<>(entries.size());
        for (Map.Entry<List<Coord>, byte[]> entry : entries) {
            List<Coord> coords = entry.getKey();
            byte[] key = new byte[coords.size()];
            for (int i = 0; i < key.length; i++) {
                key[i] = (byte) coords.get(i).index();
            }
            out.add(Map.entry(new String(key, StandardCharsets.ISO_8859_1),
                    entry.getValue().clone()));
        }
        return out;
    }

    @Override
    public int len() {
        return len;
    }

    @Override
    public void clear() {
        space.clear();
        len = 0;
    }
}
